hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/services/DefinitionsService.java

```diff
@@ -38,39 +39,108 @@
+     * manipulated in separate calls. The new <code>Trigger</code> will be persisted.  When fully defined a call to
+     * {@link #updateTrigger(String, Trigger)} is needed to enable the <code>Trigger</code>.
+     * <p>
+     * A non-group trigger can never be made into a group trigger, and vice-versa.
+     * </p>
      * @param tenantId Tenant where trigger is created
      * @param trigger New trigger definition to be added
      * @throws Exception If the <code>Trigger</code> already exists.
+     * @see {@link #addGroupTrigger(String, Trigger)} for adding a group trigger.
      */
     void addTrigger(String tenantId, Trigger trigger) throws Exception;
 
     /**
+     * Create a new Group <code>Trigger</code>. <code>Conditions</code> and <code>Dampening</code> are
+     * manipulated in separate calls. The new <code>Group Trigger</code> will be persisted.  When fully
+     * defined a call to {@link #updateGroupTrigger(String, Trigger)} is needed to enable the <code>Trigger</code>.
+     * <p>
+     * A non-group trigger can never be made into a group trigger, and vice-versa.
+     * </p>
+     * @param tenantId Tenant where trigger is created
+     * @param groupTrigger New trigger definition to be added
+     * @throws Exception If the <code>Trigger</code> already exists.
+     * @see {@link #addTrigger(String, Trigger)} for adding a non-group trigger.
+     */
+    void addGroupTrigger(String tenantId, Trigger groupTrigger) throws Exception;
+
+    /**
+     * Generate a member trigger for the specified group trigger. The dataIdMap replaces the tokens in the
+     * group trigger's conditions with actual dataIds. The member trigger gets the enabled state of the group.
+     * @param tenantId Tenant where trigger is stored
+     * @param groupId Group triggerId from which to spawn the member trigger
+     * @param memberId The member triggerId, unique id within the tenant, if null an Id will be generated
+     * @param memberName The member triggerName, not null, unique name within the tenant
+     * @param memberContext The member triggerContext. If null the context is inherited from the group trigger
+     * @param dataIdMap Tokens to be replaced in the new trigger
+     * @return the member trigger
+     * @throws Exception on any problem
+     * @see {@link #addTrigger(String, Trigger)} for adding a non-group trigger.
+     * @see {@link #addGroupTrigger(String, Trigger)} for adding a group trigger.
+     */
+    Trigger addMemberTrigger(String tenantId, String groupId, String memberId, String memberName,
+            Map<String, String> memberContext, Map<String, String> dataIdMap) throws Exception;
+
+    /**
      * The <code>Trigger</code> will be removed from the Alerts engine, as needed, and will no longer be persisted.
      * @param tenantId Tenant where trigger is stored
      * @param triggerId Trigger to be removed
+     * @throws NotFoundException if trigger is not found
```

Comment:
I see that NotFoundException has been introduced.
I have a couple of suggestions:
- If we introduce domain specific Exceptions, those should go into the API, not in the implementation. If some user wants to catch this one but not generic, it needs to explicitly declare implementation classes which should be hidden and only used on dependencies.
- Also, I see that NotFoundException has sense to be in the engine but not in the API as it has an @ApplicationException which is an implementation detail, and this annotation shouldn't go into the API.
- If we introduce a NotFoundException then it has also sense to introduce DuplicateException as the opposite when adding a new Trigger or other entity that exists.
- To make a more friendly use of the API, these domain Exceptions can be extended in the API as vanilla classes, and in the implementation extend a hidden one where the @ApplicationException annotation are set. In that way, we may have annotations in the API without explicit dependencies that are extended in the engine with specific behaviour.